### PR TITLE
chain head listener: workaround watcher deadlock

### PR DESCRIPTION
**Pull request by leoyvens**
With the logs we were able to confirm that this deadlocks within the tokio watcher implementation, this has been reported upstream https://github.com/tokio-rs/tokio/issues/4246. This PR hacks around that until it is fixed by detecting a deadlocked watcher and continuing to pull notifications, if only to discard them.

### CODE REVIEW

File: store/postgres/src/chain_head_listener.rs [hunk 1]
```diff
@@ -173,9 +178,19 @@ impl ChainHeadUpdateListener {
...
                 if let Some(watcher) = watchers.read(&logger).get(&update.network_name) {
-                    debug!(logger, "sending chain head update"; "network" => &update.network_name);
-                    watcher.send();
-                    debug!(logger, "chain head update sent"; "network" => &update.network_name);
+                    // Due to a tokio bug, we must assume that the watcher can deadlock, see
+                    // https://github.com/tokio-rs/tokio/issues/4246.
+                    if !sending_to_watcher.load(atomic::Ordering::SeqCst) {
+                        let sending_to_watcher = sending_to_watcher.cheap_clone();
+                        let sender = watcher.sender.cheap_clone();
+                        tokio::task::spawn_blocking(move || {
```
Comment (lutter) [inline store/postgres/src/chain_head_listener.rs:186]: Can that lead to a large number of dead threads hanging around? Maybe we need another `watcher_is_live: Arc<AtomicBool>` that is set to `false` just before the call to send and set to `true` right afterwards where we do not try to send if `watcher_is_live` is false.

**Review by lutter — CHANGES_REQUESTED** [review]

File: store/postgres/src/chain_head_listener.rs [hunk 2]
```diff
@@ -173,9 +178,19 @@ impl ChainHeadUpdateListener {
 
                 // If there are subscriptions for this network, notify them.
                 if let Some(watcher) = watchers.read(&logger).get(&update.network_name) {
-                    debug!(logger, "sending chain head update"; "network" => &update.network_name);
-                    watcher.send();
```
Comment (otaviopace) [inline store/postgres/src/chain_head_listener.rs:177]: We didn't `unwrap` this call before?

File: store/postgres/src/chain_head_listener.rs [hunk 3]
```diff
@@ -173,9 +178,19 @@ impl ChainHeadUpdateListener {
...
-                    debug!(logger, "chain head update sent"; "network" => &update.network_name);
+                    // Due to a tokio bug, we must assume that the watcher can deadlock, see
+                    // https://github.com/tokio-rs/tokio/issues/4246.
+                    if !sending_to_watcher.load(atomic::Ordering::SeqCst) {
+                        let sending_to_watcher = sending_to_watcher.cheap_clone();
+                        let sender = watcher.sender.cheap_clone();
+                        tokio::task::spawn_blocking(move || {
+                            sending_to_watcher.store(true, atomic::Ordering::SeqCst);
+                            sender.send(()).unwrap();
```
Comment (otaviopace) [inline store/postgres/src/chain_head_listener.rs:188]: I see that `watcher.send()` didn't receive any arguments before, why is it receiving a unit `()` value now?

Comment (leoyvens) [reply store/postgres/src/chain_head_listener.rs:188]: Before we were using the `send` on our watcher struct, which is now dead code until we can revert this hack.

File: store/postgres/src/chain_head_listener.rs (hunk 1 above)
Comment (leoyvens) [reply store/postgres/src/chain_head_listener.rs:186]: I believe the `sending_to_watcher` is already serving as the check you propose.

**Comment by leoyvens** [comment]
@lutter @otaviopace thanks for the review, comments addressed.

File: store/postgres/src/chain_head_listener.rs (hunk 1 above)
Comment (lutter) [reply store/postgres/src/chain_head_listener.rs:186]: Yes, of course, you are right.

**Review by lutter — APPROVED** [review]

**Pull request merged**